File: LeetCode/Study Plan/Practice Problems/Medium/Shortest Path In Binary Matrix/Go/shortest_path_in_binary_matrix.go
package main

import "fmt"

func is_safe(i, j, n int, grid [][]int) bool {
	return i >= 0 && j >= 0 && i < n && j < n && grid[i][j] == 0
}

var dir = [][]int{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}}

func shortest_path_binary_matrix(grid [][]int) int {
	n := len(grid)
	if grid[0][0] == 1 || grid[n-1][n-1] == 1 {
		return -1
	}

	q := [][]int{{0, 0}}
	grid[0][0] = 1
	for len(q) > 0 {
		curr := q[0]
		q = q[1:]
		i, j := curr[0], curr[1]
		dis := grid[i][j]
		if i == n-1 && j == n-1 {
			return dis
		}

		for k := 0; k < 8; k++ {
			x, y := i+dir[k][0], j+dir[k][1]
			if is_safe(x, y, n, grid) {
				q = append(q, []int{x, y})
				grid[x][y] = dis + 1
			}
		}
	}
	return -1
}

func main() {
	fmt.Println(shortest_path_binary_matrix([][]int{{0, 1}, {1, 0}}))
}

File: LeetCode/Medium/Shortest Path In Binary Matrix/Java/shortestPathInBinaryMatrix.java
class Pair {
    int x, y;
    public Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if (grid[0][0] == 1) return -1;
        int[][] dir = {{1, 0}, {1, 1}, {0, 1}, {-1, 0}, {0, -1}, {-1, -1}, {1, -1}, {-1, 1}};
        boolean[][] visited = new boolean[n][n];
        ArrayDeque<Pair> q = new ArrayDeque<>();
        q.offer(new Pair(0, 0));
        int dis = 0;
        while(!q.isEmpty()) {
            int size = q.size();
            for (int i=0; i<size; i++) {
                Pair curr = q.poll();
                if (curr.x == n-1 && curr.y == n-1) return dis+1;
                for (int k=0; k<dir.length; k++) {
                    int x = curr.x + dir[k][0];
                    int y = curr.y + dir[k][1];
                    if (x < 0 || y < 0 || x >= n || y >= n || grid[x][y] == 1 || visited[x][y]) continue;
                    q.offer(new Pair(x, y));
                    visited[x][y] = true;
                }
            }
            dis++;
        }
        return -1;
    }
}
